### DISCUSSION TITLE
WIZnet W5100S-EVB-Pico board can install micropython but after first "boot" it is not found

### DISCUSSION THREAD

**Discussion by nickehallgren** [micropython/micropython#9199] [Hardware & Peripherals]
Ok, so I have a couple of this board and one is working perfectly but the other can be programmed with micropython, I can upload a led blinking main.py but as soon as I disconnect the board from the computer it will never be found again (using Thonny or rshell as port is not found). I can also see that the main.py is not called (led not blinking) when plugged into a batterypack.

As soon as I re-flash micropython it starts again and everything works until disconnect. I've tried the nuke flash script but with the same result...

Any ideas on how to get it to work?

**Comment by robert-hh** [comment]
a) About non-working with power-pack: Are you sure that main.py was really copied to the board? Thonny does not do that by default.
b) About serial not found: Is the board directly connected to the computer or by an USB hub. In the past days I had strange phenomenon that some (most) boards connected through a HUB needed about 40 seconds to register, which with a direct connection the registered immediately. Some boards registered also fast at the HUB. That error disappeared without a reason, which is irritating. In all cases, the bootloader worked fine at the HUB.

**Reply by nickehallgren** [reply]
@robert-hh Yes it is there and starts to blink as soon as I re-install micropython. And usb is connected straight to the computer. I've tried the Arduino IDE (using it's version of blink) and there everything works as expected even after several reboots.

Even without uploading anything, the borad is not found after micropython install and disconnect. This is what I see after installation and when removed/re-inserted there is nothing...

`Bus 001 Device 019: ID 2e8a:0005 MicroPython Board in FS mode`

**Reply by robert-hh** [reply]
Do you use the same version of MicroPython on all boards?
The USB ID looks fine. Dou you see the ttyACMx deviced in /dev?
Does a plain vanilla MicroPython image work? The one for RPi Pico should start micropython, even if it does not support Ethernet.

**Reply by nickehallgren** [reply]
@robert-hh Well I started with the vanilla 1.19.1 on the other board, then I tried the one that Wiznet provided for Ethernet. But had some problems with it and finally compiled a fresh with W5100S Ethernet support and that is running very well. But on this one I've tried my own compiled and the vanilla 1.19.1 several times, also just tried the circuitpython and it too "sticks" (and works) after reboots so I don't really understand what it could be...

**Reply by nickehallgren** [reply]
@robert-hh and there is no /dev/ttyACM0 after disconnect (there when installed) or any other ACM

**Reply by robert-hh** [reply]
Dpes the tty port come up after waiting e.g. a minute? I'm asking because of the strange bug I had, where the boards registered after 40+ seconds.

**Reply by nickehallgren** [reply]
Nope, it's been connected for over 10 min now and nothing

**Comment by jimmo** [comment]
Given that it's also failing with CircuitPython that's a bad sign that there might be a hardware issue with the board (perhaps the flash chip). It's difficult though to come up with a theory that explains what you're seeing... the best I can think of is that when MicroPython writes its main.py into the flash that somehow it's causing corruption earlier in the flash.

I haven't tried this myself, but the rp2040 bootloader allows downloading of the flash via the `picotool` program (https://github.com/raspberrypi/picotool). You could use this to read back the flash from the broken one. You should find that it's identical to the image you flashed on there (up to the point where the fileystem starts).

How handy are you with gdb and SWD debugging? You can use the working board as a SWD debugger for the other one using picoprobe, openocd, and gdb?

Another option is you could try building a firmware image that never writes to flash. The easiest way to do this would be to modify `ports/rp2/modules/_boot.py` and comment the whole file out. This will prevent the board from ever formatting or mounting the flash.

You could also try modifying ports/rp2/rp2_flash.c line 58 to reduce the size that it thinks the available flash is and see if preventing it writing past a certain point helps.
`.flash_size = MICROPY_HW_FLASH_STORAGE_BYTES,`

**Reply by nickehallgren** [reply]
Maybe I was a bit unclear but it works with CircuitPython and Arduino IDE so I don't think its the flash. BTW Should I be able to see the boot.py? In micropython when listing '/' I only see my main.py and with rshell I can only list /pyboard (/ gives my computers /). Havn't done gdb and SWD

**Reply by jimmo** [reply]
> Maybe I was a bit unclear but it works with CircuitPython and Arduino IDE so I don't think its the flash.

Ah right, I thought "sticks" was bad. (But I see you've edited the post, clear now).

I'm not sure that this conclusively proves that this isn't a flash issue though -- arduino certainly not, and CircuitPython uses FAT not LittleFS which would write to different areas of the flash by default.

Actually that would be worth trying... try flashing https://micropython.org/download/ARDUINO_NANO_RP2040_CONNECT/ and see if you have the same issue. (It's a different rp2040-based board that uses FAT by default).  (Or alternatively, try compiling yourself with

`#define MICROPY_HW_USB_MSC              (1)`

added to mpconfigboard.h.  (This will enable MSC and use the FAT filesystem by default).

Make sure to use the uf2 nuke before flashing to ensure the flash is fully erased (otherwise it'll use the existing filesystem).

> BTW Should I be able to see the boot.py?

No, not unless you explicitly create one. There's a frozen `_boot.py` that I was referring to (note the leading underscore), but that won't be visible on the filesystem.

**Reply by nickehallgren** [reply]
If I understood you right I nuked, flashed the Arduino NANO RP2040 Connect. Then I used the Arduino IDE set to program Arduino Nano RP2040 Connect (with default FlashSize 16MB (no FS)) and uploaded the blink sketch. It works after disconnect and also with a batterypack

**Reply by jimmo** [reply]
No sorry. What I meant was:

- Nuke uf2
- Flash Nano Connect
- Use like a regular micropython board (i.e. copy main.py over)
- See if the same problem happens

But I've just remembered that the Nano has a much bigger flash chip than the Pico so this won't work.

You'll need to try compiling the Pico firmware (i.e. `BOARD=PICO`) yourself but with `MICROPY_HW_USB_MSC` enabled. (You can see this is already in `boards/PICO/mpconfigboard.h`, just commented out)

**Reply by nickehallgren** [reply]
I now have compiled as instructed and I can now see a 1.4MB usb drive. But when disconnect/re-connected there is nothing

**Comment by jimmo** [comment]
hi @nickehallgren -- I've been investigating this on a W5500 board.

Can you please confirm something -- are you able to reproduce this problem where the board becomes unavailable over USB if you flash a main.py that _only_ blinks an LED. i.e. does the problem only happen if your main.py tries to activate the LAN interface?

If so, could you please share the main.py that you're using?

**Reply by nickehallgren** [reply]
I'm pretty sure now that the flash is the problem. Because if I install micropython (no other code) and disconnect, it will be unavailable after connecting the usb again.

**Reply by jimmo** [reply]
@nickehallgren  Can you try what I suggested... get the board into a working state (e.g. flash nuke, power off, etc). Note also that the board seems to have enough capacitance that it has to be disconnected for a good 5+ seconds before re-connecting. Ensure that basic main.py (e.g. just flash or print) continues to work.

Then if you can get it to fail once you enable networking, can you share the code that you're using.

FWIW, I can reproduce the exact problem you're describing, but only when I misconfigure the SPI for the WizNet5k.

**Reply by nickehallgren** [reply]
I can try again (done it a few times already with just blink) when I get home from work, but I'm pretty sure the result will be a non responsive device again. As I said earlier I have another device (exactly the same) that behaves as it should...

**Reply by jimmo** [reply]
See https://github.com/micropython/micropython/pull/9224

In summary, I was able to reproduce exactly what you're describing if I attempted to construct a `network.WIZNET5K(spi)` using `spi = machine.SPI(0, baud)`, using the default pins for SPI0.

Was unable to see any problems if I either used the default `network.WIZNET5K()` (recommended), or if I fully specified the pins for `machine.SPI(0, baud, sck=..., miso=..., mosi=...)`.

**Reply by nickehallgren** [reply]
Flash nuke, disconnect for 10+ seconds, re-connect and usb drive pops up, drag vanilla 1.19.1, add blink code (below) to main.py and save to device.

```
from machine import Pin
import time

led = Pin(25, Pin.OUT)
while True:
    time.sleep_ms(500)
    led.toggle()
```
Run code, it blinks. Disconnect device, wait for 10+ seconds re-connect and no blinking, nothing, lsusb does not show anything either... So looks dead to me...
